### PR TITLE
Improvements in Resolver and some base changes for other features

### PR DESCRIPTION
**Pull request by srsuryadev**
### Summary

* Added the use of fused operators in the hustle resolver and some fixes in it.
* Added methods to set and clear the operators state. (so it can be used in the compiled query feature)
* Few base changes to support alter queries to add column to the table (added an opcode)
* Removed the old parser code, since we are using the sqlite3 parser.

### CODE REVIEW

File: src/operators/aggregate.cc [hunk 1]
```diff
@@ -570,6 +570,23 @@ void Aggregate::SortResult(std::vector<arrow::Datum>& groups,
   }
 }
 
+void Aggregate::Clear() {
+  num_aggs_ = 0;
+
+  prev_result_ = nullptr;
```
Comment (johnchronis) [inline src/operators/aggregate.cc:576]: please use the .reset() method it is more idiomatic

Also (without knowing if it makes an actual difference = nullptr and .reset in terms of operations needed. There is some casting and conversion happening when - nullptr is used
https://stackoverflow.com/questions/16151550/c11-when-clearing-shared-ptr-should-i-use-reset-or-set-to-nullptr/46577399#46577399

File: src/operators/aggregate.h [hunk 2]
```diff
@@ -132,6 +132,30 @@ class Aggregate : public BaseAggregate {
    */
   void execute(Task* ctx) override;
 
+  void Clear() override;
+
+  void set_prev_result(std::shared_ptr<OperatorResult> prev_result) {
```
Comment (johnchronis) [inline src/operators/aggregate.h:137]: You can make all of these small functions inline. A good rule is less than 10 lines should be inlined (so mutators are prime candidates)

File: src/operators/join.h [hunk 3]
```diff
@@ -67,6 +67,24 @@ class Join : public Operator {
    */
   void execute(Task *ctx) override;
 
+  void Clear() override;
+
+  void set_prev_result(std::vector<std::shared_ptr<OperatorResult>> prev_result) {
```
Comment (johnchronis) [inline src/operators/join.h:72]: same about inlining for the first 3

File: src/operators/operator.h [hunk 4]
```diff
@@ -33,10 +33,20 @@ class Operator {
  public:
   virtual void execute(Task *ctx) = 0;
 
+  virtual void Clear() = 0;
+
   std::size_t getOperatorIndex() const { return op_index_; }
 
   void setOperatorIndex(const std::size_t op_index) { op_index_ = op_index; }
 
+  void setQueryId(std::size_t query_id) {
```
Comment (johnchronis) [inline src/operators/operator.h:42]: inline?

File: src/resolver/select_resolver.h [hunk 5]
```diff
@@ -76,7 +76,6 @@ class SelectResolver {
     project_references_ =
         std::make_shared<std::vector<std::shared_ptr<ProjectReference>>>();
 
-    join_predicates_->reserve(5);
     agg_references_->reserve(5);
     group_by_references_->reserve(5);
```
Comment (johnchronis) [inline src/resolver/select_resolver.h:80]: why 5? add a comment to explain the magic number. Or better make it into a defined constant that can easily changed

File: src/storage/cmemlog.cc [hunk 6]
```diff
@@ -39,6 +39,17 @@ void memlog_add_table_mapping(int db_id, int root_page_id, char *table_name) {
...
 }
 
+void memlog_add_column_change(int db_id, int root_page_id, char* column_info) {
+  std::string new_column = std::string(column_info); 
+  std::vector<std::string> result; 
+  std::istringstream iss(new_column); 
+  std::string s;
+  while (iss >> s) { 
+    result.push_back(s);
+    std::cout << "word: " << s << std::endl;
```
Comment (johnchronis) [inline src/storage/cmemlog.cc:49]: Is this a debug message? if yes make that explicit through a comment or a debug flag.
Even better you could use something like glog to have debug and prod logs

**Review by johnchronis — APPROVED** [review]

**Review by johnchronis — APPROVED** [review]

**Comment by srsuryadev** [comment]
Thank you for the suggestions and review @johnchronis!

**Pull request merged**